### PR TITLE
refactor: HTTP client interface

### PR DESCRIPTION
**Pull request by nasdf**
## Description

This PR is split into two to make reviewing easier: #1839

This PR adds an HTTP client and server implementation that implements the `client.DB` interface and runs through the existing integration test suite.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

`make test`

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: datastore/memory/txn.go
Comment (AndrewSisley) [inline datastore/memory/txn.go:126]: For reference, for myself and other reviewers, the rest of the funcs in `basicTxn` do this (`t.closeLk`) and this appears to be a bug fix.

File: tests/integration/backup/one_to_one/import_test.go
Comment (AndrewSisley) [inline tests/integration/backup/one_to_one/import_test.go:241]: question: Why has this been added?

Comment (nasdf) [reply tests/integration/backup/one_to_one/import_test.go:241]: I believe this was due to the JSON serialization not dropping the nil field here. Potentially it could be an incorrect case within the new `resultsAreEqual` function.

Comment (nasdf) [reply tests/integration/backup/one_to_one/import_test.go]: It turns out these test case changes are no longer necessary. I've reverted this and a few other changes.

File: tests/integration/explain.go
Comment (shahzadlone) [inline tests/integration/explain.go:318]: praise+question: interesting and this sorted out what you needed?

Comment (nasdf) [reply tests/integration/explain.go:318]: Yeah we lose type info when serializing to and from JSON. This was the cleanest solution I could find for handling generic arrays.

File: tests/integration/backup/one_to_one/import_test.go
Comment (AndrewSisley) [reply tests/integration/backup/one_to_one/import_test.go]: Nice, cheers Keenan

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:68]: todo: I think the new consts would benefit from a line or two of documentation each

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:1710]: praise: cheers :laughing:

File: tests/integration/results.go
Comment (AndrewSisley) [inline tests/integration/results.go:1]: praise: Thanks for moving this to a new file, is much tidier, and I think Shahzad had been wanting this for a while :)

File: http/http_client.go
Comment (AndrewSisley) [inline http/http_client.go:59]: todo: I can't spot the reason as to why ignoring the error here is okay/desirable - can you please document this in the code with a comment if you feel like we want to do this, or handle the error.

File: http/handler_tx.go
Comment (AndrewSisley) [inline http/handler_tx.go:40]: question: I can't spot anything that 'expires' long-lived transactions which have likely been forgotten by the client.  Is that true?

I think some datastore implementations do this (not fully sure thought), meaning that the server will think a txn is alive, when the datastore has decided it is dead - are we relying on this to minimise expenses and handle errors?

question: If the above is true, are we sure we are happy tolerating a forever growing map of txns in the short-medium term?

File: http/middleware.go
Comment (AndrewSisley) [inline http/middleware.go:1]: praise: I really like how you have solved some of the complexities with Lens/Collections/txns/etc in this file :)

todo: I do think the documentation within this file needs to be expanded though, particularly on the stuff that is accessed from within the handlers (i.e. the `fooContextKey` variables).  It is not obvious when consuming them whether stuff like txns are respected.

File: http/handler_collection.go
Comment (AndrewSisley) [inline http/handler_collection.go:48]: suggestion:  I found the split here to Create/CreateMany strange, and I had to check few a few files to be sure where and why it was like this.  I understand it in the short term, but still think it is strange, and think you are kind of breaking the `client.Collection` interface here - the `http.Client` type implements the interface, but the http endpoint does not - this will be visible to users when not using the Go http-client wrapper.

I suggest you add a new ticket to change the `client.Collection` interface to remove the `CreateMany` func.  Create would then need to handle 0..n docs - I would leave it to whoever picks up that ticket how to handle that (could be an `any` type, `...Document`, or a generic constrait even).

In the new ticket please reference this code area to ensure it gets cleaned up along the way.

File: http/client.go
Comment (AndrewSisley) [inline http/client.go:87]: thought: I'm just highlighting here that our current transaction ID implementation is a bit of a hack, and exposing it more here increases the need to improve the way we manage these IDs.

File: http/client.go
Comment (AndrewSisley) [inline http/client.go:362]: todo: Same as other comment, please handle the error or document why not

File: http/client.go
Comment (AndrewSisley) [inline http/client.go:11]: question: The current (develop branch) `http` package has a playground.go file, I'm not familiar with the playground stuff at all, has this been forgotten about? Is playground.go deadcode? Have I just missed it somewhere in this PR?

**Review by AndrewSisley — COMMENTED** [review]
I really like this, but have a few questions and todos for you before I'm ready to approve.

File: tests/integration/lens.go
Comment (nasdf) [reply tests/integration/lens.go:78]: This is due to the http client losing type info when serializing to json. The newly added `assertResultsEqual` function must be used to properly compare the two values.

File: tests/integration/query/one_to_many/simple_test.go
Comment (nasdf) [reply tests/integration/query/one_to_many/simple_test.go:170]: It failed at first but now it passes with either version.

File: tests/integration/results.go
Comment (nasdf) [reply tests/integration/results.go:40]: This is due to the fact that we lose type info when serializing to json. The simple fix is to compare them using `assert.JSONEq`, but I don't think everyone would be happy with that solution.

File: tests/integration/lens.go
Comment (AndrewSisley) [reply tests/integration/lens.go:78]: You mean the http stuff is returning values of a type other than the one expected in the test(s)?

Comment (nasdf) [reply tests/integration/lens.go:78]: The underlying values are the same but the types differ slightly. 

A good example is when you serialize a number to json. It loses its type value of uint64 or float64 and becomes a string representation. When it's deserialized it is stored as an intermediary type `json.Number`. 

This allows the test suite to check the expected type and attempt to convert the `json.Number` to the correct type before comparing.

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go]: I cleaned this up a bit and changed it to only enable the `goClient` when running the change detector.

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go:73]: Added

File: http/http_client.go
Comment (nasdf) [reply http/http_client.go:62]: Added

File: tests/integration/lens.go
Comment (AndrewSisley) [reply tests/integration/lens.go:78]: Asserting the type is pretty important, particularly for the embedded Go client.  Changing the return type is a breaking change.

It also at glance looks like you are equating zero and option.None too?  This is also incorrect, including within http/gql contexts (think zero attendees vs undefined/nil - zero represents the knowledge that no one is attending, nil means that it is unknown).

File: http/client.go
Comment (nasdf) [reply http/client.go:362]: Added comment

File: http/client.go
Comment (nasdf) [reply http/client.go:336]: Added comment

File: http/middleware.go
Comment (nasdf) [reply http/middleware.go:1]: Added more documentation clarifying what each context key will be set to.

File: http/handler_tx.go
Comment (nasdf) [reply http/handler_tx.go:40]: This is a great point. I think a good long term solution would be having a shared transaction pool within the database that allows the user to specify limits and timeouts.

This will also enable us to fix the "leaky" `datastore.Txn` interface. Exposing the `MultiStore` interface on transactions is currently a pain point for the http client.

In the short-term I think this will be okay. I think only advanced users will interact with the transaction system, and we can make a note of the current short-comings somewhere.

File: http/handler_collection.go
Comment (nasdf) [reply http/handler_collection.go:48]: I had made this trade-off to keep the API more RESTful.

I do like the idea of merging the `Create` and `CreateMany` methods.

File: http/client.go
Comment (nasdf) [reply http/client.go:11]: The follow up PR addresses this. 

https://github.com/sourcenetwork/defradb/blob/19c734bc7a6da1eba2e2ee108182a14bd41f5f39/http/handler_playground.go

File: tests/integration/lens.go
Comment (nasdf) [reply tests/integration/lens.go:78]: I've updated the `assertResultsEqual` to use the new "relaxed" comparison only when the client type is not `goClientType`.

https://github.com/sourcenetwork/defradb/blob/5f96d78080d65af8b7b24e787e19f62c1b9b21c5/tests/integration/results.go#L26-L38

File: http/client.go
Comment (AndrewSisley) [reply http/client.go:336]: :) don't forget to push

File: http/handler_tx.go
Comment (AndrewSisley) [reply http/handler_tx.go:40]: Sounds sensible, sorry, I could of asked this earlier but didnt - could you please open a ticket to sort this out?

File: http/handler_collection.go
Comment (AndrewSisley) [reply http/handler_collection.go:48]: > I had made this trade-off to keep the API more RESTful.

I understood this :) Please don't forget to create that ticket though.

File: http/client.go
Comment (AndrewSisley) [reply http/client.go:11]: Cheers

File: tests/integration/lens.go
Comment (AndrewSisley) [inline tests/integration/lens.go:79]: todo: I think this comment still adds value and should be preserved somewhere.  It is not obvious reading the code as to why the order of the elements is not asserted.

File: http/client_lens.go
Comment (AndrewSisley) [inline http/client_lens.go:79]: question: Are you 100% sure this works? You appear to be unmarshalling json to an interface.  If it does work, how, and what is the resultant concrete type?  If it does not work, did the tests miss this, or did you forget to run them?  If they missed it, how?

`MigrateUp`, `MigrateDown` on `LensRegistry` and `lensHandler` appear to do this.

File: tests/integration/lens.go
Comment (AndrewSisley) [reply tests/integration/lens.go:78]: This still makes me uncomfortable:

1) The type assertion is opt-in, not opt-out, it would be very easy to accidentally not test the typing of new client interface implementations.  It would prefer that the `default` is `assert.Contains`.

2) The return types of all client interface implementations is important, and should be deterministic and consistent.  The Go client is just the most likely to be used directly by users.  That said, `http`, and later, `cli`, are new implementations, and it might be sensible to add a new ticket to the backlog to revisit this later, as to not block the main body of work in this PR.

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:192]: todo: I disagree with this default.  It will essentially double the stuff being done when running `go test` or `make test`, and will significantly degrade the dev-ex of anyone working in the codebase.

Can you please default to the go client only, and then maybe add a couple of make actions to make it easy to run http?

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:185]: todo: I think the code should change.  Whilst we currently think that running the change detector via the http client is pointless, it is not technically impossible and we shouldnt prevent it.  The environment variables should still be respected, and it will cause confusion (and bugs) if they are not.

Can you please instead change the `make test:changes` action to instead explicitly run using just the Go client.  Regardless of what the default clients are (see other comment about defaults).

File: http/client.go
Comment (AndrewSisley) [reply http/client.go:362]: Please push it :)

File: http/middleware.go
Comment (AndrewSisley) [reply http/middleware.go:1]: Looks good - thanks Keenan

File: http/client_lens.go
Comment (AndrewSisley) [reply http/client_lens.go:79]: I think I know why this was missed.

Previously we have been able to rely on the fact that we only ever had one implementation of the client interfaces, this allowed us to get away with not directly testing most functions implemented, so long as they were called internally  by functions that were directly tested.

These functions (as do most of them, our testing is horrible across the go-embedded stuff) fall into that category. They have no tests targeting them. 

I think this requires discussion at standup, as it means that significant chunks of this PR are untested, as will likely much of the CLI stuff be. There is a significant risk of other bugs besides the probable one here that I have flagged.

File: http/handler_tx.go
Comment (nasdf) [reply http/handler_tx.go:40]: Created issue #1843

Comment (AndrewSisley) [reply http/handler_tx.go:40]: Cheers :)

File: http/handler_collection.go
Comment (nasdf) [reply http/handler_collection.go:48]: Created issue #1844

Comment (AndrewSisley) [reply http/handler_collection.go:48]: Thanks :)

File: tests/integration/lens.go
Comment (nasdf) [reply tests/integration/lens.go:78]: I've updated the default comparisons like you recommended.

I'm still thinking of a solution for your second point, but it will most likely be addressed in a follow up PR.

Comment (AndrewSisley) [reply tests/integration/lens.go:78]: Cheers Keenan, and really no pressure to sort out (2) within this PR.

**Review by AndrewSisley — APPROVED** [review]
Cheers Keenan - looks really good :)  I'm approving but, would recommend you quickly check in with another dev too if any are interested as it is a significant change and they may have different opinions to mine (and catch stuff I have missed).

File: tests/integration/utils2.go
Comment (shahzadlone) [inline tests/integration/utils2.go:157]: question: what made you switch from `LookupEnv` to `Getenv`. 

IRRC `Getenv` returns an empty string, even in cases where the environment variable is not set and `LookupEnv` returns the value and a bool to indicate if it was set or not.

File: tests/integration/utils2.go
Comment (shahzadlone) [inline tests/integration/utils2.go:173]: question: What was the benefit to moving the env lookup here? I did like how they were all in one place above.

**Review by shahzadlone — COMMENTED** [review]
LGTM but I have some questions I asked and some non-blocking comments / thoughts:

1) The PR was quite easy to review, however I think it could have been even easier/clearer if:
- The testing framework changes could have been isolated in a separate PR.
- Rebase flow was used instead of the multiple merges from develop into your branch, which create those non-linear commit history with extra merge commits that aren't useful for reviewers.

2) CodeCov shows that code coverage has dropped by `-5.40%`, that is quite a lot, I am not quite sure why that is, I know sometimes it can be inaccurate and show a different number on merge, but usually it is not off by that much. This isn't a blocker but might be worth exploring / looking into it why that is.

**Comment by shahzadlone** [comment]
@nasdf I noticed that everything under `http/` has 0 coverage, is that expected for now?
image

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go:157]: We weren't using the second return value from `LookupEnv` for the boolean variables. The behavior should be the same but is a bit easier to read in my opinion.

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go:173]: I think this makes it more obvious that we have a default value for these variables. We also don't have to keep track of extra variables since the lookup return values are now scoped to the if statement.

**Comment by nasdf** [comment]
> @nasdf I noticed that everything under `http/` has 0 coverage, is that expected for now? image

I think it's not picking up the coverage from the integration tests correctly.

**Comment by AndrewSisley** [comment]
> I think it's not picking up the coverage from the integration tests correctly.

It wont do, this is expected, as the codecov job will only be running with the go client for now.  We need to come up with a performant way of handling this, for both the multi-clients, the mutations, and change detector - I think we spoke about this in a recent standup.

File: tests/integration/utils2.go
Comment (shahzadlone) [reply tests/integration/utils2.go:157]: I thought we were using the second value for some of them. However I don't mind this minor change just wanted to know the reason.

File: tests/integration/utils2.go
Comment (shahzadlone) [reply tests/integration/utils2.go:173]: I think this is subjective, I do like if all the lookups are done in the beginning and then the condition flow is executed, rather than lookup and condition checking happening at the same step. But not a strong pref, so is a non-blocking one.

**Comment by shahzadlone** [comment]
Sorry I had a small subjective nitpick I forgot to mention, regarding some function names (if these are added by you and not existent from before). I do like boolean returning functions to start with `is` or `are` to make it clear that a "question"/"equality" is being "answered"/"compared", for example:
- resultArraysAreEqual
- resultOptionsAreEqual
- resultsAreEqual
- resultsAreAnyOf


Could be:
- areResultArraysEqual
- areResultOptionsEqual
- areResultsEqual
- areResultsAnyOf

**Review by shahzadlone — APPROVED** [review]
Posted some nitpicks(non-blocking). All G otherwise

**Pull request merged**
